### PR TITLE
Tags

### PR DESCRIPTION
**Pull request by skylerwharton**
r @aharelick @riyer15 
- Created `Tag` model class. `User`s and `Resource`s can have `Tag`s.

### CODE REVIEW

File: app/models/attribute.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from .. import db
+
+users_association_table = db.Table(
+    'users_association', db.Model.metadata,
+    db.Column('tags_id', db.Integer, db.ForeignKey('tags.id')),
```
Comment (aharelick) [inline app/models/attribute.py]: I think for consistency with other foreign keys we should call this `db.Column('tag_id', ...)`

File: app/models/attribute.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    db.Column('users_id', db.Integer, db.ForeignKey('users.id'))
```
Comment (aharelick) [inline app/models/attribute.py]: `db.Column('user_id', ...)`

File: app/models/attribute.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+resources_association_table = db.Table(
+    'resources_association', db.Model.metadata,
+    db.Column('tags_id', db.Integer, db.ForeignKey('tags.id')),
```
Comment (aharelick) [inline app/models/attribute.py]: `db.Column('tag_id', ...)`

File: app/models/attribute.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    db.Column('resources_id', db.Integer, db.ForeignKey('resources.id'))
```
Comment (aharelick) [inline app/models/attribute.py]: `db.Column('resource_id', ...)`

File: app/models/attribute.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from .. import db
+
+users_association_table = db.Table(
+    'users_association', db.Model.metadata,
```
Comment (aharelick) [inline app/models/attribute.py]: I don't know if we are going to have any other many to many relationships with user, but should we name this table something like `user_tag_associations`?

File: app/models/attribute.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    db.Column('tags_id', db.Integer, db.ForeignKey('tags.id')),
+    db.Column('users_id', db.Integer, db.ForeignKey('users.id'))
+)
+
+resources_association_table = db.Table(
+    'resources_association', db.Model.metadata,
```
Comment (aharelick) [inline app/models/attribute.py]: `resource_tag_associations`?

File: app/models/attribute.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    db.Column('tags_id', db.Integer, db.ForeignKey('tags.id')),
+    db.Column('resources_id', db.Integer, db.ForeignKey('resources.id'))
+)
+
+
+class Tag(db.Model):
+    __tablename__ = 'tags'
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(30), unique=True)
```
Comment (aharelick) [inline app/models/attribute.py:19]: I'm okay with using 30 for length, just wondering if it was arbitrarily chosen or taken from somewhere?

**Comment by aharelick** [comment]
Just a few small comments if @riyer15 doesn't have any problems with it. I think we should talk more about how specifically this will be used. Just want to make sure I fully understand what you envisioned using it for.

File: app/models/attribute.py (hunk 7 above)
Comment (skylerwharton) [inline app/models/attribute.py:19]: It was arbitrarily chosen.

File: app/models/attribute.py (hunk 1 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py (hunk 2 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py (hunk 3 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py (hunk 4 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py (hunk 5 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py (hunk 6 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

**Comment by skylerwharton** [comment]
I made the suggested name changes. We can talk more about the use cases for `Tag`s on Friday.

File: app/models/attribute.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+from .. import db
+
+user_tag_associations = db.Table(
+    'users_association', db.Model.metadata,
```
Comment (aharelick) [inline app/models/attribute.py]: I was actually talking about changing the name of the table in the string `'users_association'` to `'user_tag_associations'`

File: app/models/attribute.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    db.Column('tag_id', db.Integer, db.ForeignKey('tags.id')),
+    db.Column('user_id', db.Integer, db.ForeignKey('users.id'))
+)
+
+resource_tag_associations = db.Table(
+    'resources_association', db.Model.metadata,
```
Comment (aharelick) [inline app/models/attribute.py]: same thing as above here `'resources_association'` to `'resource_tag_associations'`

File: app/models/attribute.py (hunk 8 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Ah got it - noted. In that case, I think both should be renamed. Done.

File: app/models/attribute.py (hunk 9 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done

File: app/models/attribute.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+user_tag_associations_table = db.Table(
+    'user_tag_associations', db.Model.metadata,
+    db.Column('tag_id', db.Integer, db.ForeignKey('tags.id')),
+    db.Column('user_id', db.Integer, db.ForeignKey('users.id'))
+)
+
+resource_tag_associations_table = db.Table(
+    'resource_tag_associations', db.Model.metadata,
+    db.Column('resource_tag_associations', db.Integer,
```
Comment (aharelick) [inline app/models/attribute.py]: I think you probably accidentally changed this `db.Column('resource_tag_associations',...` by accident. Should still be `db.Column('tag_id',...`

Comment (skylerwharton) [inline app/models/attribute.py]: Done.

**Comment by skylerwharton** [comment]
@aharelick @Mebsworth @riyer15 Added `Tag` subclasses :) Please suggest different names.

**Comment by skylerwharton** [comment]
r @aharelick @riyer15 

Also added methods to generate fake `AffiliationTag`s.

File: app/models/__init__.py [hunk 11]
```diff
@@ -6,3 +6,4 @@
 from user import *  # noqa
 from location import *  # noqa
 from resource import *  # noqa
+from attribute import *  # noqa
```
Comment (aharelick) [inline app/models/__init__.py:9]: the `#noqa` is actually leftover from when we copied and pasted this from something else (I don't think there's a pep8 issue). We should stop putting it there.

File: app/models/attribute.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+class Tag(db.Model):
+    __tablename__ = 'tags'
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(30), unique=True)
+    users = db.relationship('User', secondary=user_tag_associations_table,
+                            backref='tags', lazy='dynamic')
+    resources = db.relationship('Resource',
+                                secondary=resource_tag_associations_table,
+                                backref='tags', lazy='dynamic')
+    type = db.Column(db.String(50))
```
Comment (aharelick) [inline app/models/attribute.py:25]: I think `type` is already a function in python so we should change it to something else, I just saw `discriminator` in a stack overflow post (http://stackoverflow.com/questions/22602649/sqlalchemy-polymorphic-association). That might be a little intense, but let's stay away from `type`.

File: app/models/attribute.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'tag',
```
Comment (aharelick) [inline app/models/attribute.py]: do we need a `polymorphic_identity` for the parent? what happens if we remove the line?

File: app/models/attribute.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        self.name = name
+
+    @staticmethod
+    def get_by_name(name):
+        """Helper for searching by Tag name."""
+        result = Tag.query.filter_by(name=name).first()
+        return result
+
+    def __repr__(self):
+        return '<Tag \'%s\'>' % self.name
```
Comment (aharelick) [inline app/models/attribute.py]: let's print the name of the polymorphic type when we print here

File: app/models/attribute.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+
+class ResourceCategoryTag(Tag):
+    __tablename__ = 'resource_category_tags'
+    id = db.Column(db.Integer, db.ForeignKey('tags.id'), primary_key=True)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'resource_category_tag',
```
Comment (aharelick) [inline app/models/attribute.py]: `'polymorphic_identity': 'ResourceCategoryTag'` I think the identity should be the name of the class

File: app/models/attribute.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+    @staticmethod
+    def create_resource_category_tag(name):
+        """
+        Helper to create a ResourceCategoryTag entry. Returns the newly
+        created ResourceCategoryTag or the existing entry if name is already
+        in the table.
+        """
+        result = Tag.get_by_name(name)
+        if result is not None and result.type != 'resource_category_tag':
```
Comment (aharelick) [inline app/models/attribute.py]: can you throw a comment above this explaining that you cannot create a resource tag with the same name as an affiliation tag

File: app/models/attribute.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+                except ValueError:
+                    created = False
+
+
+class AffiliationTag(Tag):
+    __tablename__ = 'affiliation_tags'
+    id = db.Column(db.Integer, db.ForeignKey('tags.id'), primary_key=True)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'affiliation_tag',
```
Comment (aharelick) [inline app/models/attribute.py]: `'polymorphic_identity': 'AffiliationTag'`

File: app/models/attribute.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+    @staticmethod
+    def create_affiliation_tag(name):
+        """
+        Helper to create a AffiliationTag entry. Returns the newly
+        created AffiliationTag or the existing entry if name is already
+        in the table.
+        """
+        result = Tag.get_by_name(name)
+        if result is not None and result.type != 'affiliation_tag':
```
Comment (aharelick) [inline app/models/attribute.py]: maybe just a general comment that tag names cannot be shared across types

File: app/models/user.py [hunk 19]
```diff
@@ -215,6 +215,18 @@ def set_random_zip_codes(users, zip_codes):
         db.session.commit()
 
     @staticmethod
+    def set_random_affiliation_tags(users, affiliation_tags):
+        """Assign a random AffiliationTag from affiliation_tags to each User in
```
Comment (aharelick) [inline app/models/user.py]:
``` python
"""
Assign a random AffiliationTag from affiliation_tags to each User in
users.
"""
```

**Comment by aharelick** [comment]
very very cool stuff @weronica. Couple comments and questions, let me know if anything isn't clear.

File: app/models/__init__.py (hunk 11 above)
Comment (skylerwharton) [inline app/models/__init__.py:9]: There appears to still be a pep8 issue. For example, `'from user import *' used; unable to detect undefined names`.

Comment (aharelick) [inline app/models/__init__.py:9]: hmm, I don't see that, but I'll trust you on it. We can leave it then.

File: app/models/attribute.py (hunk 12 above)
Comment (skylerwharton) [inline app/models/attribute.py:25]: [This documentation](http://docs.sqlalchemy.org/en/rel_1_0/orm/inheritance.html) includes `type` used in this way. Thoughts?

Comment (aharelick) [inline app/models/attribute.py:25]: Hmm, I trust the documentation more than I trust me, so we should leave this as is.

File: app/models/attribute.py (hunk 15 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

File: app/models/attribute.py (hunk 16 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

File: app/models/attribute.py (hunk 17 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

File: app/models/attribute.py (hunk 18 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

File: app/models/user.py (hunk 19 above)
Comment (skylerwharton) [inline app/models/user.py]: Done.

File: app/models/attribute.py (hunk 13 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Nothing explodes when I remove it, so I think it's fine?

File: app/models/attribute.py (hunk 14 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

File: app/models/attribute.py (hunk 13 above)
Comment (aharelick) [inline app/models/attribute.py]: cool, let's get it out of there then

**Comment by skylerwharton** [comment]
I believe I have made all of the suggested changes.

File: app/models/attribute.py [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+    @staticmethod
+    def create_resource_category_tag(name):
+        """
+        Helper to create a ResourceCategoryTag entry. Returns the newly
+        created ResourceCategoryTag or the existing entry if name is already
+        in the table.
+        """
+        result = Tag.get_by_name(name)
+        # Tags must have unique names, so if a Tag that is not an
```
Comment (aharelick) [inline app/models/attribute.py]: Super Nit: `so if a Tag that is not a` vs `so if a Tag that is not an`

**Comment by aharelick** [comment]
@weronica one super nitpicky comment then feel free to merge (I trust you to make the change and merge).

File: app/models/attribute.py (hunk 20 above)
Comment (skylerwharton) [inline app/models/attribute.py]: Done.

**Pull request merged**